File: GFG/Heap/ksorted.cpp
#include<iostream>
#include"../ds.h"
using namespace std;


void ksort(vector<int> &arr,int k){
    MinHeap<int> q;
    for(int i=0;i<=k;i++){
        q.insert(arr[i]);
    }
    int ind=0;
    for(int j=k+1;j<arr.size();j++){
        arr[ind++]=q.heap[0];
        q.insert(arr[j]);
    }
    ind=0;
    while(q.isEmpty()==false){
        arr[ind++]=q.heap[0];
        q.extractMin();
    }
    return;
}

int main(){
    vector<int> arr{9,8,7,18,19,17};
    // vector<int> arr{10,9,7,8,4,70,50,60};
    int k=2;
    ksort(arr,k);
    for(auto i:arr){
        cout<<i<<" ";
    }
    return 0;
}

File: GFG/Heap/ksorted.py
import ds


def ksort(arr,k):
    q=ds.MinHeap()
    for i in range(0,k+1):
        q.insert(arr[i])
    ind=0
    for i in range(k+1,len(arr)):
        
        print(ind)
        arr[ind]=q.heap[0]
        ind+=1
        q.insert(arr[i])
    ind=0
    print(q.heap)
    while(q.is_empty()==False):
        
        print(ind)
        arr[ind]=q.heap[0]
        ind+=1
        print(q.heap)
        q.extract_min()
        
    

arr=[9,8,7,18,19,17]
ksort(arr,2)
print(arr)

